### PR TITLE
Automatically login & reauth

### PR DESCRIPTION
**Pull request by leoleoasd**
Discussed in #1690

- [x] This pr contains no AI generated code
- [x] cargo fmt has been run
- [x] cargo clippy has been run
- [x] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: tools/cli/src/cli/common.rs [hunk 1]
```diff
@@ -73,33 +83,37 @@ impl CommonOpt {
...
+            Some(_username) => {
+                username = _username.clone();
                 // Is it in the store?
-                match tokens.get(username) {
+                match tokens.get(&username) {
                     Some(t) => t.clone(),
                     None => {
                         error!("No valid authentication tokens found for {}.", username);
-                        std::process::exit(1);
+                        return Err(ToClientError::Other);
```
Comment (yaleman) [inline tools/cli/src/cli/common.rs:116]: We know the username at this point so this shouldn't be "other"

File: tools/cli/src/cli/common.rs [hunk 2]
```diff
@@ -158,22 +176,81 @@ impl CommonOpt {
...
+    }
+
+    #[async_recursion]
+    pub async fn to_client(&self, optype: OpType) -> KanidmClient {
+        match self.try_to_client(optype.clone()).await {
+            Ok(c) => c,
+            Err(e) => {
+                match e {
+                    ToClientError::NeedLogin(username) => {
+                        let login_opt = Select::with_theme(&ColorfulTheme::default())
```
Comment (yaleman) [inline tools/cli/src/cli/common.rs:204]: This should just be a [dialoguer::Confirm](https://docs.rs/dialoguer/0.10.4/dialoguer/struct.Confirm.html)

File: tools/cli/src/cli/common.rs [hunk 3]
```diff
@@ -158,22 +176,81 @@ impl CommonOpt {
...
+                            password: env::var("KANIDM_PASSWORD").ok(),
+                        };
+                        login_opt.exec().await;
+                        // we still use `to_client` instead of `try_to_client` because we may need to prompt user to re-auth again.
+                        // since reauth_opt will call `to_client`, this function is recursive anyway.
+                        // we use copt since it's username is updated.
+                        return login_opt.copt.to_client(optype).await;
+                    }
+                    ToClientError::NeedReauth(username) => {
+                        let login_opt = Select::with_theme(&ColorfulTheme::default())
```
Comment (yaleman) [inline tools/cli/src/cli/common.rs:227]: Same as above, should be a [dialoguer::Confirm](https://docs.rs/dialoguer/0.10.4/dialoguer/struct.Confirm.html)

**Review by yaleman — CHANGES_REQUESTED** [review]
A few minor changes please, looking good so far!

**Review by Firstyear — APPROVED** [review]
I think this looks really good, it's a good idea and improvement to make things nicer for users. Thanks!

**Review by yaleman — APPROVED** [review]
Awesome, thanks!

**Pull request merged**